src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+            return GetEquivalenceKey(preference, isRemovableAssignment);
+        }
+
+        protected abstract SyntaxNode UpdateNameForFlaggedNode(SyntaxNode node, SyntaxToken newName);
+        protected abstract SyntaxToken GetForEachStatementIdentifier(TForEachStatementSyntax node);
+        protected abstract TBlockSyntax GenerateBlock(IEnumerable<TStatementSyntax> statements);
+        protected abstract ILocalSymbol GetSingleDeclaredLocal(TLocalDeclarationStatementSyntax localDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken);
+        protected abstract void InsertAtStartOfSwitchCaseBlock(TSwitchCaseBlockSyntax switchCaseBlock, SyntaxEditor editor, TLocalDeclarationStatementSyntax declarationStatement);
+
+        private bool NeedsToMoveNewLocalDeclarationsNearReference(string diagnosticId) => diagnosticId == IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId;
+        
+
+        protected override bool IncludeDiagnosticDuringFixAll(FixAllState fixAllState, Diagnostic diagnostic)
+        {
+            return fixAllState.CodeActionEquivalenceKey == GetEquivalenceKey(diagnostic) &&
+                !IsForEachIterationVariableDiagnostic(diagnostic, fixAllState.Document);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root,
+            out string diagnosticId,
+            out UnusedValuePreference preference,
+            out bool removeAssignments)
+        {
+            diagnosticId = diagnostics[0].Id;
+            preference = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetUnusedValuePreference(diagnostics[0]);
+            removeAssignments = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostics[0]);
+#if DEBUG
+            foreach (var diagnostic in diagnostics)
+            {
+                Debug.Assert(diagnosticId == diagnostic.Id);
+                Debug.Assert(preference == AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetUnusedValuePreference(diagnostic));
+                Debug.Assert(removeAssignments == AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic));
+            }
+#endif
+
+            return GetDiagnosticsGroupedByMember(diagnostics, syntaxFacts, root);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root)
+            => diagnostics.GroupBy(d => syntaxFacts.GetContainingMemberDeclaration(root, d.Location.SourceSpan.Start));
+
+        protected override async Task<Document> FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
```

Comment:
we're a subclass of SyntaxEditorBaedCodeFixProvider.  But you're not using the SyntaxEditor entrypoints... it's unclear what's going on here.

Reply:
This is an override of the `FixAllAsync` overload _before_ the SyntaxEditor is instantiated by `SyntaxEditorBasedCodeFixProvider`. We are updating the underlying document to have annotated original member declaration nodes (i.e. all method bodies) on which at least one diagnostic was reported, so that `PostProcessDocumentAsync` can operate on these. SyntaxEditor is then generated for the annotated document by the `base.FixAllAsync` invocation at the end of this override.

---
In reply to: [230604277](https://github.com/dotnet/roslyn/pull/30777#discussion_r230604277) [](ancestors = 230604277)

Reply:
sorry, i still don't actually see why that's necessary.  you are passed in everything you need to FixAllAsync, including the Document.  You are free to do whatever you need there, *including making a new editor*, and doing the work there (including post processing).  Then, just get the root back you need, and call `.ReplaceNode(root, newRoot)` on the original editor you got back. 

Does that make sense?